p2p/key.go

```diff
@@ -0,0 +1,122 @@
+package p2p
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+
+	crypto "github.com/tendermint/go-crypto"
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+type ID string
+
+//------------------------------------------------------------------------------
+// Persistent peer ID
+// TODO: encrypt on disk
+
+// NodeKey is the persistent peer key.
+// It contains the nodes private key for authentication.
+type NodeKey struct {
+	PrivKey crypto.PrivKey `json:"priv_key"` // our priv key
+}
+
+// ID returns the peer's canonical ID - the hash of its public key.
+func (nodeKey *NodeKey) ID() ID {
+	return ID(hex.EncodeToString(nodeKey.id()))
+}
+
+func (nodeKey *NodeKey) id() []byte {
+	return nodeKey.PrivKey.PubKey().Address()
+}
+
+// PubKey returns the peer's PubKey
+func (nodeKey *NodeKey) PubKey() crypto.PubKey {
+	return nodeKey.PrivKey.PubKey()
+}
+
+func (nodeKey *NodeKey) SatisfiesTarget(target []byte) bool {
+	return bytes.Compare(nodeKey.id(), target) < 0
+}
+
+// LoadOrGenNodeKey attempts to load the NodeKey from the given filePath,
+// and checks that the corresponding ID is less than the target.
+// If the file does not exist, it generates and saves a new NodeKey
+// with ID less than target.
+func LoadOrGenNodeKey(filePath string, target []byte) (*NodeKey, error) {
+	if cmn.FileExists(filePath) {
+		nodeKey, err := loadNodeKey(filePath)
+		if err != nil {
+			return nil, err
+		}
+		if !nodeKey.SatisfiesTarget(target) {
+			return nil, fmt.Errorf("Loaded ID (%s) does not satisfy target (%X)", nodeKey.ID(), target)
+		}
+		return nodeKey, nil
+	} else {
+		return genNodeKey(filePath, target)
+	}
+}
+
+// MakePoWTarget returns a 20 byte target byte array.
+func MakePoWTarget(difficulty uint8) []byte {
+	zeroPrefixLen := (int(difficulty) / 8)
+	prefix := bytes.Repeat([]byte{0}, zeroPrefixLen)
+	mod := (difficulty % 8)
+	if mod > 0 {
+		nonZeroPrefix := byte(1 << (8 - mod))
+		prefix = append(prefix, nonZeroPrefix)
+	}
+	return append(prefix, bytes.Repeat([]byte{255}, 20-len(prefix))...)
```

Comment:
Could we use `0xFF` instead?
```go
bytes.Repeat([]byte{0xFF}, 20-len(prefix
```

Also this code is going to crash starting at a difficulty of 161 until the end because:
161/8 = 20
161%8 = 1
so len(prefix) = 21
20 - 21 = -1 which finna crash bytes.Repeat